cipd_packages/device_doctor/lib/src/android_device.dart

```diff
@@ -380,7 +381,54 @@ class AndroidDevice implements Device {
 
   @override
   Future<void> recover() async {
-    await eval('adb', <String>['-s', deviceId!, 'reboot'], canFail: false);
+    await cleanDevice();
+  }
+
+  @visibleForTesting
+  Future<void> waitForDeviceUp({ProcessManager? processManager}) async {
+    processManager ??= LocalProcessManager();
+    final List<String> waitForDeviceArgs = [
+      'shell',
+      'true',
+    ];
+    final int maxTries = 4;
+    final int sleepSecs = 5;
+    for (var i = 0; i < maxTries; i++) {
+      final Process proc = await startProcess('adb', waitForDeviceArgs, processManager: processManager);
+      if (0 == await proc.exitCode) {
+        break;
+      }
+      await Future.delayed(Duration(seconds: sleepSecs));
+    }
+  }
+
+  @visibleForTesting
+  Future<bool> cleanDevice({ProcessManager? processManager}) async {
+    processManager ??= LocalProcessManager();
+    // Ensure device is rebooted after deleting cache and packages.
+    final List<String> deletePackageCacheArgs = [
+      'shell pm list packages -f',
+      '|',
```

Comment:
I'm pretty sure you can't create pipes like this with dart:io. What I'd do is something like:

```dart
  final io.ProcessResult result = io.Process.runSync('adb', <String>['shell', 'pm', 'list', 'packages', '-f']);
  if (result.exitCode != 0) {
    handleError(result.stderr as String);
  }
  // Or whatever the dart equivalent of your awk script would be
  final String pid = (result.stdout as String).split('=')[2];
  // or whatever the equivalent of your xargs would be...
  io.Process.runSync('adb', <String>['shell', 'pm', 'clear', pid]);
```